sklearn/utils/metadata_requests.py

```diff
@@ -0,0 +1,735 @@
+    """Get a MetadataRequest instance from the given object.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    obj : object
+        If the object is already a MetadataRequest, return that.
+        If the object is an estimator, try to call `get_metadata_request` and get
+        an instance from that method.
+        If the object is a dict, create a MetadataRequest from that.
+
+    Returns
+    -------
+    metadata_requests : MetadataRequest
+        A ``MetadataRequest`` taken or created from the given object.
+    """
+    if obj is None:
+        return MetadataRequest()
+
+    if isinstance(obj, MetadataRequest):
+        return obj
+
+    if isinstance(obj, dict):
+        return MetadataRequest(obj)
+
+    try:
+        return obj.get_metadata_request(output="MetadataRequest")
+    except TypeError:
+        # The method exists, but doesn't accept `output`
+        return MetadataRequest(obj.get_metadata_request())
+    except AttributeError:
+        # The object doesn't have a `get_metadata_request` method.
+        return MetadataRequest()
+
+
+class MetadataRouter:
+    """Route the metadata to child objects.
+
+    .. versionadded:: 1.1
+    """
+
+    def __init__(self):
+        self.requests = MetadataRequest()
+
+    def add(self, *obj, mapping="one-to-one", overwrite=False, mask=False):
+        """Add a set of requests to the existing ones.
+
+        Parameters
+        ----------
+        *obj : objects
+            A set of objects from which the requests are extracted. Passed as
+            arguments to this method.
+
+        mapping : dict or str, default="one-to-one"
+            The mapping between the ``obj``'s methods and this routing object's
+            methods. If ``"one-to-one"`` all methods' requests from ``obj`` are
+            merged into this instance's methods. If a dict, the mapping is of
+            the form ``{"destination_method": "source_method"}``.
+
+        overwrite : bool or str, default=False
+            - True: ``alias`` replaces the existing routing.
+            - False: a ``ValueError`` is raised if the given value conflicts
+              with an existing one.
+            - "on-default": only overwrite the alias if it is
+              RequestType.ERROR_IF_PASSED
+
+        mask : bool, default=False
+            If the requested metadata should be masked by the alias. If
+            ``True``, then a request of the form
+            ``{'sample_weight' : 'my_weight'}`` is converted to
+            ``{'my_weight': 'my_weight'}``. This is required for meta-estimators
+            which should expose the requested parameters and not the ones
+            expected by the objects' methods.
+        """
+        for x in obj:
+            if mask:
+                x = metadata_request_factory(x).masked()
+            self.requests.add_requests(x, mapping=mapping, overwrite=overwrite)
+        return self
+
+    def get_metadata_request(self, output="dict"):
```

Comment:
I don't see what we gain by allowing output to be switched as a parameter here... Would it be sufficient to suggest in the documentation that the output of `get_metadata_request` (returning a dict) can be wrapped in `MetadataRequest` for a useful API? Or should we always return a MetadataRequest?

Reply:
yeah, wanted to have the `dict` as the default in case third parties don't want to have the `MetadataRequest` object. I changed it to always returning `dict`.